Alert maintainers on failure to run upgrade-provider

When the upgrade-provider tool fails to run, we should alert maintainers that automatic upgrades may be failing.

Creating a GitHub Issue would be appropriate, if we could ensure that we only update/replace a particular open issue by name to avoid spamming a repository.

### Background

Pulumi Azure has had a series of [upgrade provider failures](https://github.com/pulumi/pulumi-azure/actions/workflows/upgrade-provider.yml), some stemming from a [dependency issue in the bridge](https://github.com/pulumi/pulumi-azure/actions/runs/5252151683/jobs/9487857739) or due to errors following attempted automatic token mapping:

```
Falling back to direct application...
panic: fatal: error An assertion has failed: failed to compute default tokens. source error: 2 errors occurred:
        * resources:
3 errors occurred:
        * could not find a module that prefixes 'arc_machine_extension'
```

--- a/main.go
+++ b/main.go
@@ -274,8 +274,8 @@ func createFailureIssue(ctx upgrade.Context, repoOrg string, repoName string) (s
 		"--limit=1",
 		"--json=url",
 	)
-	bytes1 := new(bytes.Buffer)
-	getLatestWorkflowRun.Stdout = bytes1
+	ghRunBytes := new(bytes.Buffer)
+	getLatestWorkflowRun.Stdout = ghRunBytes
 	err := getLatestWorkflowRun.Run()
 	if err != nil {
 		return "createFailureIssue error: failed to query latest workflow run", err
@@ -283,7 +283,7 @@ func createFailureIssue(ctx upgrade.Context, repoOrg string, repoName string) (s
 	var workflowRunUrls []struct {
 		Url string `json:"url"`
 	}
-	err = json.Unmarshal(bytes1.Bytes(), &workflowRunUrls)
+	err = json.Unmarshal(ghRunBytes.Bytes(), &workflowRunUrls)
 	if err != nil {
 		return "createFailureIssue error: failed to unmarshal `gh run list` output", err
 	}
@@ -299,7 +299,8 @@ func createFailureIssue(ctx upgrade.Context, repoOrg string, repoName string) (s
 		"--state=open",
 		"--author=pulumi-bot",
 	)
-	getIssues.Stdout = new(bytes.Buffer)
+	ghSearchBytes := new(bytes.Buffer)
+	getIssues.Stdout = ghSearchBytes
 	err = getIssues.Run()
 	if err != nil {
 		return "createFailureIssue error: failed to search existing issues", err
@@ -308,7 +309,7 @@ func createFailureIssue(ctx upgrade.Context, repoOrg string, repoName string) (s
 		Title  string `json:"title"`
 		Number int    `json:"number"`
 	}{}
-	err = json.Unmarshal(getIssues.Stdout.Bytes(), &titles)
+	err = json.Unmarshal(ghSearchBytes.Bytes(), &titles)
 	if err != nil {
 		return "createFailureIssue error: failed to unmarshal `gh search issues` output", err
 	}

--- a/main.go
+++ b/main.go
@@ -142,7 +142,7 @@ func cmd() *cobra.Command {
 			err := upgrade.UpgradeProvider(context, repoOrg, repoName)
 			if err != nil {
 				msg, err := createFailureIssue(context, repoOrg, repoName)
-				if err != nil {
+				if err != nil && context.InferVersion {
 					fmt.Println(msg)
 				}
 			}

--- a/main.go
+++ b/main.go
@@ -140,9 +140,9 @@ func cmd() *cobra.Command {
 		},
 		Run: func(_ *cobra.Command, args []string) {
 			err := upgrade.UpgradeProvider(context, repoOrg, repoName)
-			if err != nil {
+			if err != nil && context.InferVersion {
 				msg, err := createFailureIssue(context, repoOrg, repoName)
-				if err != nil && context.InferVersion {
+				if err != nil {
 					fmt.Println(msg)
 				}
 			}

--- a/main.go
+++ b/main.go
@@ -141,7 +141,7 @@ func cmd() *cobra.Command {
 		Run: func(_ *cobra.Command, args []string) {
 			err := upgrade.UpgradeProvider(context, repoOrg, repoName)
 			if err != nil {
-				msg, err := createFailureIssue(context, repoOrg, repoName, err.Error())
+				msg, err := createFailureIssue(context, repoOrg, repoName)
 				if err != nil {
 					fmt.Println(msg)
 				}
@@ -258,12 +258,15 @@ func bindFlags(cmd *cobra.Command, v *viper.Viper) {
 }
 
 // Create an issue in the provider repo with a message describing the upgrade failure
-func createFailureIssue(ctx upgrade.Context, repoOrg string, repoName string, errMsg string) (string, error) {
+func createFailureIssue(ctx upgrade.Context, repoOrg string, repoName string) (string, error) {
 	now := time.Now()
-	h, min, _ := now.Clock()
-	y, m, d := now.Date()
-	title := fmt.Sprintf("Upgrade provider failure: %v %v, %v @ %v:%v", m, d, y, h, min)
-	errMsg = "Error: \n`" + errMsg + "`"
+	hr, min, _ := now.Clock()
+	yr, mth, day := now.Date()
+	minStr := fmt.Sprintf("%v", min)
+	if min < 10 {
+		minStr = fmt.Sprintf("0%v", min)
+	}
+	title := fmt.Sprintf("Upgrade provider failure: %v %v, %v @ %v:%v", mth, day, yr, hr, minStr)
 
 	getLatestWorkflowRun := exec.CommandContext(ctx, "gh", "run", "list",
 		"--workflow="+"upgrade-provider.yml",
@@ -314,7 +317,7 @@ func createFailureIssue(ctx upgrade.Context, repoOrg string, repoName string, er
 		cmd := exec.Command(
 			"gh", "issue", "create",
 			"--repo="+repoOrg+"/"+repoName,
-			"--body="+fmt.Sprintf("%s\n\nView the latest workflow run: %s", errMsg, workflowUrl),
+			"--body="+fmt.Sprintf("View the latest workflow run: %s", workflowUrl),
 			"--title="+title,
 			"--label="+"needs-triage",
 		)
@@ -329,7 +332,7 @@ func createFailureIssue(ctx upgrade.Context, repoOrg string, repoName string, er
 		cmd := exec.Command("gh", "issue", "comment",
 			fmt.Sprintf("%v", t.Number),
 			"--repo="+repoOrg+"/"+repoName,
-			"--body="+fmt.Sprintf("%s\n%s\n\nView the latest workflow run: %s", title, errMsg, workflowUrl),
+			"--body="+fmt.Sprintf("%s\n\nView the latest workflow run: %s", title, workflowUrl),
 		)
 		if err := cmd.Run(); err != nil {
 			return "createFailureIssue error: Failed to update existing issue", err

--- a/main.go
+++ b/main.go
@@ -280,9 +280,9 @@ func createFailureIssue(ctx upgrade.Context, repoOrg string, repoName string) (s
 	if err != nil {
 		return "createFailureIssue error: failed to query latest workflow run", err
 	}
-	workflowRunUrls := []struct {
+	var workflowRunUrls []struct {
 		Url string `json:"url"`
-	}{}
+	}
 	err = json.Unmarshal(bytes1.Bytes(), &workflowRunUrls)
 	if err != nil {
 		return "createFailureIssue error: failed to unmarshal `gh run list` output", err
@@ -296,9 +296,9 @@ func createFailureIssue(ctx upgrade.Context, repoOrg string, repoName string) (s
 		"Upgrade provider failure: ",
 		"--repo="+repoOrg+"/"+repoName,
 		"--json=title,number",
+		"--state=open",
 	)
-	bytes2 := new(bytes.Buffer)
-	getIssues.Stdout = bytes2
+	getIssues.Stdout = new(bytes.Buffer)
 	err = getIssues.Run()
 	if err != nil {
 		return "createFailureIssue error: failed to search existing issues", err
@@ -307,7 +307,7 @@ func createFailureIssue(ctx upgrade.Context, repoOrg string, repoName string) (s
 		Title  string `json:"title"`
 		Number int    `json:"number"`
 	}{}
-	err = json.Unmarshal(bytes2.Bytes(), &titles)
+	err = json.Unmarshal(getIssues.Stdout.Bytes(), &titles)
 	if err != nil {
 		return "createFailureIssue error: failed to unmarshal `gh search issues` output", err
 	}
@@ -317,7 +317,7 @@ func createFailureIssue(ctx upgrade.Context, repoOrg string, repoName string) (s
 		cmd := exec.Command(
 			"gh", "issue", "create",
 			"--repo="+repoOrg+"/"+repoName,
-			"--body="+fmt.Sprintf("View the latest workflow run: %s", workflowUrl),
+			"--body=View the latest workflow run: "+workflowUrl,
 			"--title="+title,
 			"--label="+"needs-triage",
 		)

--- a/main.go
+++ b/main.go
@@ -297,6 +297,7 @@ func createFailureIssue(ctx upgrade.Context, repoOrg string, repoName string) (s
 		"--repo="+repoOrg+"/"+repoName,
 		"--json=title,number",
 		"--state=open",
+		"--author=pulumi-bot",
 	)
 	getIssues.Stdout = new(bytes.Buffer)
 	err = getIssues.Run()

--- a/main.go
+++ b/main.go
@@ -260,48 +260,80 @@ func bindFlags(cmd *cobra.Command, v *viper.Viper) {
 // Create an issue in the provider repo with a message describing the upgrade failure
 func createFailureIssue(ctx upgrade.Context, repoOrg string, repoName string, errMsg string) (string, error) {
 	now := time.Now()
+	h, min, _ := now.Clock()
 	y, m, d := now.Date()
-	title := fmt.Sprintf("Upgrade provider failure: %v %v %v", y, m, d)
+	title := fmt.Sprintf("Upgrade provider failure: %v %v, %v @ %v:%v", m, d, y, h, min)
+	errMsg = "Error: \n`" + errMsg + "`"
+
+	getLatestWorkflowRun := exec.CommandContext(ctx, "gh", "run", "list",
+		"--workflow="+"upgrade-provider.yml",
+		"--repo="+repoOrg+"/"+repoName,
+		"--limit=1",
+		"--json=url",
+	)
+	bytes1 := new(bytes.Buffer)
+	getLatestWorkflowRun.Stdout = bytes1
+	err := getLatestWorkflowRun.Run()
+	if err != nil {
+		return "createFailureIssue error: failed to query latest workflow run", err
+	}
+	workflowRunUrls := []struct {
+		Url string `json:"url"`
+	}{}
+	err = json.Unmarshal(bytes1.Bytes(), &workflowRunUrls)
+	if err != nil {
+		return "createFailureIssue error: failed to unmarshal `gh run list` output", err
+	}
+	if len(workflowRunUrls) == 0 {
+		return "createFailureIssue error: no workflow runs detected", err
+	}
+	workflowUrl := workflowRunUrls[0].Url
 
 	getIssues := exec.CommandContext(ctx, "gh", "search", "issues",
 		"Upgrade provider failure: ",
 		"--repo="+repoOrg+"/"+repoName,
 		"--json=title,number",
 	)
-	bytes := new(bytes.Buffer)
-	getIssues.Stdout = bytes
-	err := getIssues.Run()
+	bytes2 := new(bytes.Buffer)
+	getIssues.Stdout = bytes2
+	err = getIssues.Run()
 	if err != nil {
-		return "Failed to create failure issue: failed to search existing issues", err
+		return "createFailureIssue error: failed to search existing issues", err
 	}
 	titles := []struct {
 		Title  string `json:"title"`
 		Number int    `json:"number"`
 	}{}
-	err = json.Unmarshal(bytes.Bytes(), &titles)
+	err = json.Unmarshal(bytes2.Bytes(), &titles)
 	if err != nil {
-		return "Failed to create failure issue: failed to unmarshal gh output", err
+		return "createFailureIssue error: failed to unmarshal `gh search issues` output", err
 	}
-	for _, t := range titles {
-		cmd := exec.Command("gh", "issue", "delete",
-			fmt.Sprintf("%v", t.Number),
+
+	// create new issue if none exist
+	if len(titles) == 0 {
+		cmd := exec.Command(
+			"gh", "issue", "create",
 			"--repo="+repoOrg+"/"+repoName,
-			"--yes",
+			"--body="+fmt.Sprintf("%s\n\nView the latest workflow run: %s", errMsg, workflowUrl),
+			"--title="+title,
+			"--label="+"needs-triage",
 		)
 		if err := cmd.Run(); err != nil {
-			return "Failed to create failure issue: failed to delete old failure issues", err
+			return "createFailureIssue error: failed to create new issue", err
+		}
+
+		return "", err
+	}
+	// otherwise, comment on existing issue(s) with a link to the latest workflow ID
+	for _, t := range titles {
+		cmd := exec.Command("gh", "issue", "comment",
+			fmt.Sprintf("%v", t.Number),
+			"--repo="+repoOrg+"/"+repoName,
+			"--body="+fmt.Sprintf("%s\n%s\n\nView the latest workflow run: %s", title, errMsg, workflowUrl),
+		)
+		if err := cmd.Run(); err != nil {
+			return "createFailureIssue error: Failed to update existing issue", err
 		}
 	}
-
-	cmd := exec.Command(
-		"gh", "issue", "create",
-		"--repo="+repoOrg+"/"+repoName,
-		"--body="+fmt.Sprintf("%q", errMsg),
-		"--title="+fmt.Sprintf("%q", title),
-	)
-	if err := cmd.Run(); err != nil {
-		return "Failed to create failure issue: failed to create new issue", err
-	}
-
 	return "", err
 }

--- a/main.go
+++ b/main.go
@@ -1,12 +1,16 @@
 package main
 
 import (
+	"bytes"
 	"context"
+	"encoding/json"
 	"errors"
 	"fmt"
 	"go/build"
 	"os"
+	"os/exec"
 	"strings"
+	"time"
 
 	semver "github.com/Masterminds/semver/v3"
 	"github.com/pulumi/pulumi/sdk/v3/go/common/util/contract"
@@ -136,6 +140,12 @@ func cmd() *cobra.Command {
 		},
 		Run: func(_ *cobra.Command, args []string) {
 			err := upgrade.UpgradeProvider(context, repoOrg, repoName)
+			if err != nil {
+				msg, err := createFailureIssue(context, repoOrg, repoName, err.Error())
+				if err != nil {
+					fmt.Println(msg)
+				}
+			}
 			exitOnError(err)
 		},
 	}
@@ -245,4 +255,53 @@ func bindFlags(cmd *cobra.Command, v *viper.Viper) {
 			contract.AssertNoErrorf(err, "error setting flag")
 		}
 	})
+}
+
+// Create an issue in the provider repo with a message describing the upgrade failure
+func createFailureIssue(ctx upgrade.Context, repoOrg string, repoName string, errMsg string) (string, error) {
+	now := time.Now()
+	y, m, d := now.Date()
+	title := fmt.Sprintf("Upgrade provider failure: %v %v %v", y, m, d)
+
+	getIssues := exec.CommandContext(ctx, "gh", "search", "issues",
+		"Upgrade provider failure: ",
+		"--repo="+repoOrg+"/"+repoName,
+		"--json=title,number",
+	)
+	bytes := new(bytes.Buffer)
+	getIssues.Stdout = bytes
+	err := getIssues.Run()
+	if err != nil {
+		return "Failed to create failure issue: failed to search existing issues", err
+	}
+	titles := []struct {
+		Title  string `json:"title"`
+		Number int    `json:"number"`
+	}{}
+	err = json.Unmarshal(bytes.Bytes(), &titles)
+	if err != nil {
+		return "Failed to create failure issue: failed to unmarshal gh output", err
+	}
+	for _, t := range titles {
+		cmd := exec.Command("gh", "issue", "delete",
+			fmt.Sprintf("%v", t.Number),
+			"--repo="+repoOrg+"/"+repoName,
+			"--yes",
+		)
+		if err := cmd.Run(); err != nil {
+			return "Failed to create failure issue: failed to delete old failure issues", err
+		}
+	}
+
+	cmd := exec.Command(
+		"gh", "issue", "create",
+		"--repo="+repoOrg+"/"+repoName,
+		"--body="+fmt.Sprintf("%q", errMsg),
+		"--title="+fmt.Sprintf("%q", title),
+	)
+	if err := cmd.Run(); err != nil {
+		return "Failed to create failure issue: failed to create new issue", err
+	}
+
+	return "", err
 }

